File: quant1x/id/id64/id.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

import base64
from dataclasses import dataclass

PAYLOAD_BITS = 22


@dataclass(frozen=True)
class ID:
    """64 位可排序标识（8 字节 BigEndian）。

    位布局（动态位宽）：

    | 1bit 符号(恒 0) | Physical(41bit, epoch 相对毫秒) | NodeID(workerBits) | Seq(seqBits) |

    workerBits / seqBits 由节点总数推导（见 ``with_node_count``），
    因此 ``node_id`` / ``seq`` 解析需要传入对应的 ``worker_bits``。
    """

    raw: bytes

    def __post_init__(self) -> None:
        if len(self.raw) != 8:
            raise ValueError("ID expects exactly 8 bytes")

    @classmethod
    def from_int(cls, value: int) -> "ID":
        return cls(value.to_bytes(8, "big", signed=False))

    @classmethod
    def from_bytes(cls, raw: bytes) -> "ID":
        return cls(raw)

    def bytes(self) -> bytes:
        return self.raw

    def to_int(self) -> int:
        return int.from_bytes(self.raw, "big", signed=False)

    def __str__(self) -> str:
        return base64.urlsafe_b64encode(self.raw).rstrip(b"=").decode("ascii")

    def physical(self) -> int:
        """返回 epoch 相对毫秒（高 41 位）。"""
        return self.to_int() >> PAYLOAD_BITS

    def node_id(self, worker_bits: int) -> int:
        shift = PAYLOAD_BITS - worker_bits
        return (self.to_int() >> shift) & ((1 << worker_bits) - 1)

    def seq(self, worker_bits: int) -> int:
        shift = PAYLOAD_BITS - worker_bits
        return self.to_int() & ((1 << shift) - 1)

File: quant1x/id/id64/Id.java
// Copyright (c) 2026 Quant1X. All rights reserved.
// Author: wangfeng <wangfengxy@sina.cn>
// SPDX-License-Identifier: MIT

package quant1x.id.id64;

import java.util.Arrays;
import java.util.Base64;
import java.util.Objects;

/**
 * 64 位可排序标识。
 *
 * <p>位布局（动态位宽，与 Go/Python 版一致）：
 * <pre>
 * | 1bit 符号(恒 0) | Physical(41bit, epoch 相对毫秒) | NodeID(workerBits) | Seq(seqBits) |
 * </pre>
 *
 * <p>{@link #nodeId(int)} / {@link #seq(int)} 解析需要传入生成器对应的 workerBits。
 */
public final class Id {

    private static final int RAW_LENGTH = 8;

    private final byte[] raw;

    private Id(byte[] raw) {
        if (raw == null || raw.length != RAW_LENGTH) {
            throw new IllegalArgumentException("id64: Id expects exactly 8 bytes");
        }
        this.raw = raw;
    }

    /** 从 long 位模式构造 */
    public static Id fromLong(long value) {
        byte[] raw = new byte[RAW_LENGTH];
        ByteIO.writeUint64(raw, 0, value);
        return new Id(raw);
    }

    /** 从 BigEndian 8 字节构造 */
    public static Id fromBytes(byte[] raw) {
        return new Id(raw);
    }

    /** BigEndian 8 字节表示 */
    public byte[] bytes() {
        return raw.clone();
    }

    /** long 位模式 */
    public long toLong() {
        return ByteIO.readUint64(raw, 0);
    }

    /** base64url 无填充字符串（8 字节 → 11 字符） */
    @Override
    public String toString() {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(raw);
    }

    /** 返回 epoch 相对毫秒（高 41 位） */
    public long physical() {
        return toLong() >> HLC.PAYLOAD_BITS;
    }

    /** 返回节点标识，workerBits 必须与生成器配置一致 */
    public long nodeId(int workerBits) {
        int shift = HLC.PAYLOAD_BITS - workerBits;
        return (toLong() >>> shift) & ((1L << workerBits) - 1);
    }

    /** 返回序列号，workerBits 必须与生成器配置一致 */
    public long seq(int workerBits) {
        int shift = HLC.PAYLOAD_BITS - workerBits;
        return toLong() & ((1L << shift) - 1);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof Id)) {
            return false;
        }
        return Arrays.equals(raw, ((Id) obj).raw);
    }

    @Override
    public int hashCode() {
        return Objects.hash(Arrays.hashCode(raw));
    }
}
